### PR TITLE
Handle `import`ing relocated dispatch functions

### PR DESCRIPTION
**Pull request by jakirkham**
Requires PR ( https://github.com/rapidsai/cudf/pull/8342 )

As these functions recently got relocated, handle `import`ing from the new location with a fallback to the old location.

xref: https://github.com/dask/dask/pull/7503
xref: https://github.com/dask/dask/pull/7505

### CODE REVIEW

File: dask_cuda/proxy_object.py
Comment (galipremsagar) [inline dask_cuda/proxy_object.py:26]: `group_split_dispatch` and some others too seems to be present in `dask.dataframe.core` in old and new version of code. Should we just import it from there?

Comment (jakirkham) [reply dask_cuda/proxy_object.py]: Good point. Updated below using modules that appear to work in both cases for all of these

**Review by galipremsagar — APPROVED** [review]

**Review by quasiben — APPROVED** [review]

**Comment by quasiben** [comment]
Thanks @jakirkham for the work and @galipremsagar for the review!

**Comment by quasiben** [comment]
@gpucibot merge

**Comment by jakirkham** [comment]
Seeing this on [CI]( https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-v0.20/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.0,GPU_LABEL=gpu,OS=ubuntu16.04,PYTHON=3.7/95/console )

```python
19:52:07 ________________________ test_dataframes_share_dev_mem _________________________
19:52:07 
19:52:07     def test_dataframes_share_dev_mem():
19:52:07         cudf = pytest.importorskip("cudf")
19:52:07     
19:52:07         df = cudf.DataFrame({"a": range(10)})
19:52:07 >       grouped = shuffle_group(df, "a", 0, 2, 2, False, 2)
19:52:07 
19:52:07 dask_cuda/tests/test_proxify_host_file.py:109: 
19:52:07 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
19:52:07 dask_cuda/proxify_device_objects.py:142: in wrapper
19:52:07     ret = func(*args, **kwargs)
19:52:07 /opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/shuffle.py:816: in shuffle_group
19:52:07     ind = hash_object_dispatch(df[cols] if cols else df, index=False)
19:52:07 /opt/conda/envs/rapids/lib/python3.7/site-packages/dask/utils.py:511: in __call__
19:52:07     meth = self.dispatch(type(arg))
19:52:07 /opt/conda/envs/rapids/lib/python3.7/site-packages/dask/utils.py:499: in dispatch
19:52:07     return self.dispatch(cls)  # recurse
19:52:07 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
19:52:07 
19:52:07 self = <dask.utils.Dispatch object at 0x7fe9f6b57fd0>
19:52:07 cls = <class 'cudf.core.dataframe.DataFrame'>
19:52:07 
19:52:07     def dispatch(self, cls):
19:52:07         """Return the function implementation for the given ``cls``"""
19:52:07         # Fast path with direct lookup on cls
19:52:07         lk = self._lookup
19:52:07         try:
19:52:07             impl = lk[cls]
19:52:07         except KeyError:
19:52:07             pass
19:52:07         else:
19:52:07             return impl
19:52:07         # Is a lazy registration function present?
19:52:07         toplevel, _, _ = cls.__module__.partition(".")
19:52:07         try:
19:52:07             register = self._lazy.pop(toplevel)
19:52:07         except KeyError:
19:52:07             pass
19:52:07         else:
19:52:07             register()
19:52:07             return self.dispatch(cls)  # recurse
19:52:07         # Walk the MRO and cache the lookup result
19:52:07         for cls2 in inspect.getmro(cls)[1:]:
19:52:07             if cls2 in lk:
19:52:07                 lk[cls] = lk[cls2]
19:52:07                 return lk[cls2]
19:52:07 >       raise TypeError("No dispatch for {0}".format(cls))
19:52:07 E       TypeError: No dispatch for <class 'cudf.core.dataframe.DataFrame'>
19:52:07 
19:52:07 /opt/conda/envs/rapids/lib/python3.7/site-packages/dask/utils.py:505: TypeError
```

**Comment by jakirkham** [comment]
cc @madsbk

**Comment by galipremsagar** [comment]
This error is because `hash_object_dispatch` location has changed and `dask-cudf` code is not yet updated with correct location. dask-cudf will silently fail as the import is in try/catch and the dispatch is never registered for `cudf.DataFrame`. Fix for this is already included in : https://github.com/rapidsai/cudf/pull/8342/files#diff-8cccf94d2261588b89ab7cd02573abfe37a19aab98fb6d210184c5ce293e9b3eR249

**Comment by galipremsagar** [comment]
rerun tests

**Comment by jakirkham** [comment]
rerun tests

(looks like packages are up now)

**Comment by kkraus14** [comment]
Looks like there's a failure: https://gpuci.gpuopenanalytics.com/job/rapidsai/job/gpuci-v0.20/job/dask-cuda/job/prb/job/dask-cuda-gpu-test/CUDA=11.2,GPU_LABEL=gpu,OS=centos7,PYTHON=3.8/104/testReport/junit/dask_cuda.tests/test_ucx_options/test_global_option/

**Comment by jakirkham** [comment]
Yeah saw that. Though it's not related to this change. Looks like some UCX configuration test failure. My inclination is to mark it as xfail so we can get this out and unblock people on CI who still need these changes. Thoughts?

File: dask_cuda/tests/test_ucx_options.py
Comment (jakirkham) [inline dask_cuda/tests/test_ucx_options.py:56]: Ok marked this test with `xfail` so we can get these changes out-the-door to fix other CI issues. Though raised issue ( https://github.com/rapidsai/dask-cuda/issues/627 ) to track/investigate

Comment (quasiben) [reply dask_cuda/tests/test_ucx_options.py:56]: I think this is ok.  @pentschev do you have any concerns?  we are still in burn down so if something does crop up we have time to handle

Comment (jakirkham) [reply dask_cuda/tests/test_ucx_options.py:56]: For clarity cuDF's and RAFT's CI still fails without these changes, which is blocking a fair number of people. No objection to revisiting, but I do think we need to do this short term.

Comment (pentschev) [reply dask_cuda/tests/test_ucx_options.py:56]: This is weird, it never failed before, including in my nightly runs with both [UCX 1.9](https://github.com/pentschev/ucx-py-ci/issues/64) and [UCX master](https://github.com/pentschev/ucx-py-ci/issues/65). I'm actually concerned that this may be some error from a previous test that pytest is spilling over. I opened a PR to trigger CI in https://github.com/rapidsai/dask-cuda/pull/628 , let's hold on and see if that fails too before merging this.

Comment (jakirkham) [reply dask_cuda/tests/test_ucx_options.py:56]: Agree it is weird. I don't recall seeing this error on this PR before that build either.

Ok though JFYI before this came up, Ben had used a merge command upthread ( https://github.com/rapidsai/dask-cuda/pull/623#issuecomment-847439931 ). So the bot may wind up merging this anyways if passes.

**Comment by kkraus14** [comment]
I think this needs to be updated as well: https://github.com/rapidsai/cudf/blob/branch-21.06/conda/recipes/custreamz/meta.yaml#L34-L35

File: dask_cuda/tests/test_ucx_options.py
Comment (pentschev) [reply dask_cuda/tests/test_ucx_options.py:56]: Haha, yeah, it was too quick, so no worries. I'm just now wondering if this isn't a bit dangerous, one could do any changes to the PR after someone told the bot to merge, and unless somebody sees it in time, it would be merged.

Comment (jakirkham) [reply dask_cuda/tests/test_ucx_options.py:56]: Yeah it probably should be associated with a commit hash

Comment (pentschev) [reply dask_cuda/tests/test_ucx_options.py:56]: Or just invalidated based on timestamp. If commit timestamp > gpucibot merge comment, invalidate automerge.

Comment (pentschev) [reply dask_cuda/tests/test_ucx_options.py:56]: I mean the time the commit was pushed, not the local commit timestamp.

Comment (jakirkham) [reply dask_cuda/tests/test_ucx_options.py:56]: Submitted PR ( https://github.com/rapidsai/dask-cuda/pull/630 ) to revert this piece. Probably still fails, but at least that gives us something to work from

**Pull request merged**
